### PR TITLE
[RFC] State scripts morty

### PR DESCRIPTION
**Pull request by drewmoseley**
This is a backport of the client-side changes needed to support state scripts in the Mender client.

This is not yet ready for merge.  Being submitted for automated testing only.

### CODE REVIEW

**Comment by drewmoseley** [comment]
@kacf FYI.  Please take a look.  What JIRA issue(s) should I connect this with?

**Comment by drewmoseley** [comment]
@kacf I created https://tracker.mender.io/browse/MEN-1326 for this work.  I'm not sure where to link that into the planning epics and such.

**Comment by drewmoseley** [comment]
@kacf the failed test seems to be due to:
```
ERROR: Layer 'mender' depends on layer 'go', but this layer is not enabled in your configuration
```
Is there something that has changed to make morty builds fail?  I know we used the external go layer for that branch and it is no longer needed for pyro/master branch builds.

**Comment by kacf** [comment]
> @kacf I created https://tracker.mender.io/browse/MEN-1326 for this work. I'm not sure where to link that into the planning epics and such.

That is a question for @eysteins.

> `ERROR: Layer 'mender' depends on layer 'go', but this layer is not enabled in your configuration`

Yes, you need to clone and add the https://github.com/mem/oe-meta-go layer for morty builds.

**Comment by drewmoseley** [comment]
@kacf understood.  I have the oe-meta-go layer in my local morty builds but it doesn't appear to be there in the Jenkins based builds.  Where does that need to be added?

**Comment by kacf** [comment]
To be honest I thought it was there already: https://github.com/mendersoftware/mender-qa/blob/master/scripts/jenkins-yoctobuild-build.sh#L138-L142

Maybe check the logs that this line is really hit?

**Comment by drewmoseley** [comment]
@kacf I see where the test is happening in the log but it seems that the poky sources are not the right revision.  It seems the default of "master" is used for POKY_REV. Should that be conditional based on the target branch of this PR instead?  Is that even possible?

**Comment by drewmoseley** [comment]
OK. It appears the version of mender-artifact being used (2.0) doesn't properly handle the state script inclusion.   I have the following in _tmp/work/vexpress_qemu-poky-linux-gnueabi/core-image-full-cmdline/1.0-r0/temp/log.do_image_mender_
```
+ mender-artifact write rootfs-image -n target-image-1.0 -t vexpress-qemu -s /work/dmoseley/Mender/yocto/morty/build/vexpress-qemu/tmp/deploy/images/vexpress-qemu/mender-state-scripts -u /work/dmoseley/Mender/yocto/morty/build/vexpress-qemu/tmp/work/vexpress_qemu-poky-linux-gnueabi/core-image-full-cmdline/1.0-r0/deploy-core-image-full-cmdline-image-complete/core-image-full-cmdline-vexpress-qemu.ext4 -o /work/dmoseley/Mender/yocto/morty/build/vexpress-qemu/tmp/work/vexpress_qemu-poky-linux-gnueabi/core-image-full-cmdline/1.0-r0/deploy-core-image-full-cmdline-image-complete/core-image-full-cmdline-vexpress-qemu-20170713205121.rootfs.mender
Incorrect Usage: flag provided but not defined: -s
```

Looking through the git sources it seems that version 2.0.x should have proper state script support but there do appear to be some changes in the master branch that may be relevant.  Especially https://github.com/mendersoftware/mender-artifact/commit/8c1e09e7e75b271fd7dc0b1ff2b7271b1f94cc30

Should mender-artifact v2.0.x work here?

**Comment by drewmoseley** [comment]
OK.  I see that the '-s' parameter was explicitly removed from the mender-artifact tool for the 2.0.x branch here: https://github.com/mendersoftware/mender-artifact/commit/abfba0408c1a8298c8a3d917ddb5434bede9b636

I think this means that we need to use the _git version of the recipe.  I did manually confirm that if I modify the branch and commit has then this works.

**Comment by drewmoseley** [comment]
On a related note, I added https://tracker.mender.io/browse/MEN-1328 to get better error handling in case the mender-artifact utility fails.  Right now it seems to report an error which is shown in the log file buried in tmp/work/\<blah\> but bitbake itself continues with no indication of an issue.

**Comment by drewmoseley** [comment]
OK.  I added a commit with the mender and mender-artifact git recipes.  These are locked to the commit hashes as of today and don't have all the dynamic autorev stuff but that should be ok for the purposes of this.  I have not actually tested the runtime yet but the mender artifacts are now properly created.

**Comment by kacf** [comment]
Yep, the dynamic autorev stuff doesn't work properly on morty, I remember this from when I first added it. It's only a convenience though, the hardcoded stuff will work just fine.

**Comment by kacf** [comment]
> I see where the test is happening in the log but it seems that the poky sources are not the right revision. It seems the default of "master" is used for POKY_REV. Should that be conditional based on the target branch of this PR instead? Is that even possible?

We have some autodetection in place here, but I doubt it will work for morty. You're probably better off running it manually with the right branch in https://ci.cfengine.com/job/yoctobuild-pr/.

**Comment by drewmoseley** [comment]
I added a new commit (meta-mender-demo/recipes-mender/example-state-scripts/files/example-script) to add some additional logging from the state scripts.  I verified that I am running the git version of the Mender client and when it connects to the git version of the Mender integration server I was expecting to see some of the logging data in /data when I deployed an artifact but I do not.

**Comment by drewmoseley** [comment]
FYI, this is not yet ready for prime time.  Still trying to get the exact versions of mender and mender-artifact worked out.

In reality we probably do not want to merge this to morty as it has not been tested against non-state-script configs.  We probably will want to push this to https://github.com/mendersoftware/meta-mender as a new branch.

File: meta-mender-core/recipes-mender/mender/mender_git.bb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 require mender.inc
 
-SRC_URI = "git://github.com/mendersoftware/mender;protocol=https;branch=master"
+SRC_URI = "git://github.com/pasinskim/mender;protocol=https;branch=state_scripts"
```
Comment (kacf) [inline meta-mender-core/recipes-mender/mender/mender_git.bb]: This can be changed back to master now, it's been merged.

**Comment by drewmoseley** [comment]
@kacf this has been updated to the proper versions of mender and mender-artifact from the master branch.  It is ready for final review.

File: meta-mender-demo/recipes-mender/example-state-scripts/files/example-script [hunk 2]
```diff
@@ -1,3 +1,4 @@
 #!/bin/sh
 
+echo "$(cat /etc/mender/artifact_info): $(basename "$0") was called!" >> /data/mender-state-script-debug.txt
```
Comment (kacf) [inline meta-mender-demo/recipes-mender/example-state-scripts/files/example-script]: You might want to add this to master as well!

**Comment by drewmoseley** [comment]
OK.  I'll push this as a new branch into our github repo.  Closing.

**Pull request closed without merging**